### PR TITLE
[WTF] Ensure isMainThread semantics for Linux match Apple/BSD semantics

### PR DESCRIPTION
**Pull request by jameshilliard**
#### 50d7bfe855ca043cc72ec09b3d3d75100015101b
<pre>
[WTF] Ensure isMainThread semantics for Linux match Apple/BSD semantics
<a href="https://bugs.webkit.org/show_bug.cgi?id=242690">https://bugs.webkit.org/show_bug.cgi?id=242690</a>

Reviewed by Darin Adler.

It&apos;s not entirely clear if the current implementation semantics are fully
identical between Linux and Apple/BSD systems. The Apple/BSD pthread_main_np
function has 3 potential return values, 1 if the thread is the initial thread,
0 if it is not or -1 if thread is not initialized/unknown, lets make sure we
only return true for isMainThread if the return for pthread_main_np is 1.

On linux checking if we are on the initial thread can be accomplished by
checking the equality of getpid and gettid. Note we use a syscall for calling
gettid as the gettid() libc function may not be guaranteed to be supported by
all libc&apos;s on Linux unlike the syscall(which is also used elsewhere in WebKit).

* Source/WTF/wtf/generic/MainThreadGeneric.cpp:
(WTF::initializeMainThreadPlatform):
(WTF::isMainThread):

Canonical link: <a href="https://commits.webkit.org/252485@main">https://commits.webkit.org/252485@main</a>
</pre>

### CODE REVIEW

File: Source/WTF/wtf/generic/MainThreadGeneric.cpp [hunk 1]
```diff
@@ -31,30 +31,39 @@
...
+#if !HAVE(PTHREAD_MAIN_NP) && !OS(LINUX)
     mainThread = pthread_self();
 #endif
 }
 
 bool isMainThread()
 {
 #if HAVE(PTHREAD_MAIN_NP)
-    return pthread_main_np();
+    return pthread_main_np() == 1;
```
Comment (darinadler) [inline Source/WTF/wtf/generic/MainThreadGeneric.cpp]: The rest of this patch changes behavior on Linux. This part attempts to fix a bug on the `HAVE(PTHREAD_MAIN_NP)` platforms. I’d like to understand how we’re testing this part.

Comment (jameshilliard) [reply Source/WTF/wtf/generic/MainThreadGeneric.cpp]: I guess we can do something like this to make sure this call isn't failing:
```cpp
    int mainThreadNp = pthread_main_np();
    ASSERT(mainThreadNp != -1);
    return mainThreadNp == 1;
```

**Review by darinadler — APPROVED** [review]

**Comment by Constellation** [comment]
This is nice! I always wonder whether we can have a reliable good way to do this on Windows too.

**Comment by Constellation** [comment]
The best way I can find so far is, enumerating all threads and compare creation time. The smallest thread is the thread starting the process.

**Comment by Constellation** [comment]
But, yeah, it is ugly...

**Comment by jameshilliard** [comment]
> There is not a good justification for changing it without a test or evidence.

I'm just going off the docs in regards to the return code behavior, the `-1` error condition seems hard(maybe impossible?) to hit so not sure how best to test it other than an assert. Comparing against 1 seemed more correct though at least based on the docs I found, hard to say for sure though.

**Comment by jameshilliard** [comment]
> I always wonder whether we can have a reliable good way to do this on Windows too.

Hmm, maybe there is an equivalent to the Linux approach for windows using: [`GetCurrentThreadId()`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentthreadid) [`GetCurrentProcessId()`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentprocessid).

**Comment by jameshilliard** [comment]
> The smallest thread is the thread starting the process.

Does windows differentiates pthread id's with kernel thread id's(gettid/getpid on Linux for example are both kernel thread id's that can be compared with each other, those differ from the pthread id's, on Linux you can't directly compare kernel thread id's against pthread id's)?

**Comment by jameshilliard** [comment]
Looks like using the Apple/BSD `pthread_main_np()` semantics is actually wrong, see #3497.

**Pull request merged**